cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -798,21 +859,87 @@ func isRollingUpdateFinished(md *v1alpha1.MachineDeployment) bool {
 
 // getMachineDeploymentUntilDeadline returns error only when fetching the machineDeployment has been failing consequently and deadline is crossed
 func (m *McmManager) getMachineDeploymentUntilDeadline(mdName string, retryInterval time.Duration, deadline time.Time) (*v1alpha1.MachineDeployment, error) {
+	ctx, cancelFn := context.WithDeadline(context.Background(), deadline)
+	defer cancelFn()
 	for {
 		md, err := m.machineDeploymentLister.MachineDeployments(m.namespace).Get(mdName)
-		if err != nil && time.Now().Before(deadline) {
-			klog.Warningf("Unable to fetch MachineDeployment object %s, Error: %s, will retry in %s", mdName, err, retryInterval)
-			time.Sleep(conflictRetryInterval)
-			continue
-		} else if err != nil {
-			// Timeout occurred
-			klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %s, timeout occurred", mdName, err)
-			return nil, err
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %+v, timeout occurred", mdName, err)
+				return nil, err
+			case <-time.After(retryInterval):
+				klog.Warningf("Unable to fetch MachineDeployment object %s, Error: %+v, will retry the operation", mdName, err)
+				continue
+			}
 		}
 		return md, nil
 	}
 }
 
+// getMachineUntilDeadline returns error only when fetching the machine has been failing consequently and deadline is crossed
+func (m *McmManager) getMachineUntilDeadline(mcName string, retryInterval time.Duration, deadline time.Time) (*v1alpha1.Machine, error) {
```

Comment:
we are fetching `machine(s)` and `node` from cache... which always returns fast , so we shouldn't have a `waitUnitlDeadline` for these.

Reply:
we have a `getMachineDeploymentUntilDeadline` that has a `waitUntilDeadline` for machine deployments. It also fetches it from the cache.

Reply:
removed deadlines for cache calls in [9b7f40d](https://github.com/gardener/autoscaler/pull/215/commits/9b7f40d4fa9dd8371045b4f766f3b62b9b659182)